### PR TITLE
Sockets - Kelly and Pauline

### PR DESCRIPTION
**Pull request by kdow**
# Adagrams
Congratulations! You're submitting your assignment.

## Comprehension Questions

Feature | Feedback
--- | ---
What are the components that make up a method? | The components that make up a method are: the name of the method, arguments (values that are passed), code block, and return value.
What are the advantages of using git when collaboratively working on one code base? | To track changes and easily work on different computers by pulling the code that was pushed.
What kind of relationship did you and your pair have with the unit tests? | We had a good relationship with the unit tests. We found them useful for knowing what needed more work and when to move on. 
Does your code use any methods from the `Enumerable` mixin? If so, where and why was it helpful? | Yes, we used max_by and min_by. We used max_by to find the word with the highest score, and min_by to find the shortest word in case of a tie.
What was one method you and your pair used to debug code? | We used VS Code's debugger to track variables, types, etc.
What are two discussion points that you and your pair discussed when giving/receiving feedback from each other that you would be willing to share? | We found it helpful to have our pair to keep us on track when one of us was mentally drained, the other would keep things moving. We also found it helpful to have our partner to come up with ideas for debugging.

### CODE REVIEW

File: lib/adagrams.rb [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+end
+
+def highest_score_from(words)
+  scored_words = []
+  highest_scored_word = 0
+  words.each do |word|
+    scored_words << {:word => word, :score => score_word(word)}
+  end
+
+  highest_scored_word = scored_words.max_by do |scored_words|
```
Comment (dHelmgren) [inline lib/adagrams.rb:99]: `scored_words` here does a weird thing to your `scored_words` from line 93. Make sure to check your error messages before you submit your code! https://en.wikipedia.org/wiki/Variable_shadowing

Comment (dHelmgren) [reply lib/adagrams.rb:99]: You could easily avoid this by calling the line thusly:

```ruby
highest_scored_word = scored_words.max_by do |word|
```

File: lib/adagrams.rb [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+  highest_score = highest_scored_word[:score]
+  highest_score_array = []
+  scored_words.each do |word|
+    if word[:score] == highest_score
+      highest_score_array.push(word)
+    end
+  end
+
+  if highest_score_array.length > 1
+    highest_scored_word = highest_score_array.min_by { |words| words[:word].length }
```
Comment (dHelmgren) [inline lib/adagrams.rb:112]: again, this `words` variable is shadowing something. This one not causing any harm, but it CAN do really weird and bad stuff. In the same scope, try to use unique names.

File: lib/adagrams.rb [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+  letter_pool = make_letter_pool(letter_pool, "B", 2)
+  letter_pool = make_letter_pool(letter_pool, "C", 2)
+  letter_pool = make_letter_pool(letter_pool, "D", 4)
+  letter_pool = make_letter_pool(letter_pool, "E", 12)
+  letter_pool = make_letter_pool(letter_pool, "F", 2)
+  letter_pool = make_letter_pool(letter_pool, "G", 3)
+  letter_pool = make_letter_pool(letter_pool, "H", 2)
+  letter_pool = make_letter_pool(letter_pool, "I", 9)
+  letter_pool = make_letter_pool(letter_pool, "J", 1)
+  letter_pool = make_letter_pool(letter_pool, "K", 1)
```
Comment (dHelmgren) [inline lib/adagrams.rb:24]: I don't think this formatting makes this any easier to read. This takes up more space than a loop would have, and more space than hard-coding an array would have.

File: lib/adagrams.rb [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+  letter_pool = make_letter_pool(letter_pool, "Y", 2)
+  letter_pool = make_letter_pool(letter_pool, "Z", 1)
+
+  available_letters = []
+  available_letters = letter_pool.dup
+
+  10.times do
+    curr_letter = letter_pool.sample
+    user_hand.push(curr_letter)
+    available_letters.delete(curr_letter)
```
Comment (dHelmgren) [inline lib/adagrams.rb:47]: This letter generation method will never result in duplicate letters, because the `.delete` method deletes _all_ of the `curr_letter` in the array.

File: lib/adagrams.rb [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+  end
+  return user_hand
+end
+
+def uses_available_letters?(input, letters_in_hand)
+  input_letters = input.chars
+  input_letters.each do |letter|
+    if letters_in_hand.include?(letter)
+      letters_in_hand.delete(letter)
```
Comment (dHelmgren) [inline lib/adagrams.rb:56]: This never got caught because of draw letters, but this code would break for sets with duplicate letters, because `.delete` removes duplicates.

**Comment by dHelmgren** [comment]
# Adagrams
## What We're Looking For

Feature | Feedback
--- | ---
**General** |
Answered comprehension questions | yes
Both teammates contributed to the codebase | Yes
Small commits with meaningful commit messages | Yes
**Code Requirements** |
`draw_letters` method | See comment!
Uses appropriate data structure to store the letter distribution | Yes!
All tests for `draw_letters` pass | Yes
`uses_available_letters?` method | See Comment!
All tests for `uses_available_letters?` pass | Yes
`score_word` method | Yes
Uses appropriate data structure to store the letter scores | Yeah
All tests for `score_word` pass | Yes
`highest_score_from` method | Yes
Appropriately handles edge cases for tie-breaking logic | Yes
All tests for `highest_score_from` pass |Yes
**Overall** | Hey y'all! Great work on this project. Your code fulfills all of the requirements and does it in a readable and reasonable way. The code is clean, and it passes all the tests! Take a look at the comments, as there are a few sneaky bugs that got through the tests.
